sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -103,17 +102,19 @@ case class ScalaUDF(
     }
   }
 
-  private def createToScalaConverter(i: Int, dataType: DataType): Any => Any = {
-    if (inputEncoders.isEmpty) {
-      // for untyped Scala UDF
-      CatalystTypeConverters.createToScalaConverter(dataType)
+  private def scalaConverter(i: Int, dataType: DataType): Any => Any = {
+    if (inputEncoders.isEmpty || // for untyped Scala UDF
+        inputEncoders(i).isEmpty || // for types aren't supported by encoder, e.g. Any
+        inputPrimitives(i) || // inputPrimitives is not empty when inputEncoders is not empty
+        dataType.existsRecursively(_.isInstanceOf[UserDefinedType[_]])) {
+      createToScalaConverter(dataType)
     } else {
-      val encoder = inputEncoders(i)
-      if (encoder.isDefined && encoder.get.isSerializedAsStructForTopLevel) {
-        val fromRow = encoder.get.resolveAndBind().createDeserializer()
+      val enc = inputEncoders(i).get
+      val fromRow = enc.createDeserializer()
+      if (enc.isSerializedAsStructForTopLevel) {
         row: Any => fromRow(row.asInstanceOf[InternalRow])
       } else {
-        CatalystTypeConverters.createToScalaConverter(dataType)
+        value: Any => fromRow(new GenericInternalRow(Array(value)))
```

Comment:
it's too expensive to create a `GenericInternalRow` and `Array` for each input. We should do
```
val inputRow = new GenericInternalRow(1)
value: Any => inputRow.set(0, value); fromRow(inputRow)
```